homeassistant/components/motioneye/__init__.py

```diff
@@ -0,0 +1,263 @@
+    )
+
+    async_dispatcher_send(
+        hass,
+        SIGNAL_CAMERA_ADD.format(entry.entry_id),
+        camera,
+    )
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up motionEye from a config entry."""
+    client = create_motioneye_client(
+        entry.data[CONF_URL],
+        admin_username=entry.data.get(CONF_ADMIN_USERNAME),
+        admin_password=entry.data.get(CONF_ADMIN_PASSWORD),
+        surveillance_username=entry.data.get(CONF_SURVEILLANCE_USERNAME),
+        surveillance_password=entry.data.get(CONF_SURVEILLANCE_PASSWORD),
+    )
+
+    try:
+        await client.async_client_login()
+    except MotionEyeClientInvalidAuthError:
+        await client.async_client_close()
+        await _create_reauth_flow(hass, entry)
+        return False
+    except MotionEyeClientError as exc:
+        await client.async_client_close()
+        raise ConfigEntryNotReady from exc
+
+    async def async_update_data() -> dict[str, Any] | None:
+        try:
+            return await client.async_get_cameras()
+        except MotionEyeClientError as exc:
+            raise UpdateFailed("Error communicating with API") from exc
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=DOMAIN,
+        update_method=async_update_data,
+        update_interval=DEFAULT_SCAN_INTERVAL,
+    )
+    hass.data[DOMAIN][entry.entry_id] = {
+        CONF_CLIENT: client,
+        CONF_COORDINATOR: coordinator,
+        CONF_ON_UNLOAD: [],
+    }
+
+    current_cameras: set[str] = set()
+    device_registry = await dr.async_get_registry(hass)
+
+    def _async_process_motioneye_cameras() -> None:
+        """Process motionEye camera additions and removals."""
+        inbound_camera: set[str] = set()
+        if KEY_CAMERAS not in coordinator.data:
+            return
+
+        for camera in coordinator.data[KEY_CAMERAS]:
+            if not is_acceptable_camera(camera):
+                return
+            camera_id = camera[KEY_ID]
+            device_unique_id = get_motioneye_device_unique_id(entry.entry_id, camera_id)
+            inbound_camera.add(device_unique_id)
+
+            if device_unique_id in current_cameras:
+                continue
+            current_cameras.add(device_unique_id)
+            hass.async_create_task(
+                _add_camera(
+                    hass,
+                    device_registry,
+                    client,
+                    entry,
+                    camera_id,
+                    camera,
+                    device_unique_id,
+                )
+            )
+
+        # Ensure every device associated with this config entry is still in the list of
+        # motionEye cameras, otherwise remove the device (and thus entities).
+        for device_entry in dr.async_entries_for_config_entry(
+            device_registry, entry.entry_id
+        ):
+            for (kind, key) in device_entry.identifiers:
```

Comment:
Although suggested, it's not currently enforced that each identifier is a tuple of 2 values. So this can fail.

Reply:
Hmm. That's actually better, as I need to store the config entry id *and* a (non-unique) camera_id, and need to use both separately in future PRs. Rather than merge into a string, and separate them later, this is an opportunity to just store the independent values in the tuple (i.e. make it a tuple of 3, if that is not verboten).

Reply:
That's fine.